hazelcast/cluster.py

```diff
@@ -1,63 +1,116 @@
-from hazelcast.lifecycle import LIFECYCLE_STATE_CONNECTED, LIFECYCLE_STATE_DISCONNECTED
-from hazelcast.protocol.codec import client_add_membership_listener_codec, client_authentication_codec
-from hazelcast.util import get_possible_addresses, get_provider_addresses, calculate_version
-from hazelcast.version import CLIENT_TYPE, CLIENT_VERSION, SERIALIZATION_VERSION
+from hazelcast import six
+from hazelcast.errors import TargetDisconnectedError, IllegalStateError
+from hazelcast.util import check_not_none
 
-# Membership Event Types
-MEMBER_ADDED = 1
-MEMBER_REMOVED = 2
 
+class _MemberListSnapshot(object):
+    __slots__ = ("version", "members")
 
-class ClusterService(object):
+    def __init__(self, version, members):
+        self.version = version
+        self.members = members
+
+
+class ClientInfo(object):
+    """
+    Local information of the client.
     """
-    Hazelcast cluster service. It provides access to the members in the cluster and the client can register for changes
-    in the cluster members.
 
-    All the methods on the Cluster are thread-safe.
+    __slots__ = ("uuid", "address", "name", "labels")
+
+    def __init__(self, client_uuid, address, name, labels):
+        self.uuid = client_uuid
+        """Unique id of this client instance."""
+
+        self.address = address
+        """Local address that is used to communicate with cluster."""
+
+        self.name = name
+        """Name of the client."""
+
+        self.labels = labels
+        """Read-only set of all labels of this client."""
+
+    def __repr__(self):
+        return "ClientInfo(uuid=%s, address=%s, name=%s, labels=%s)" % (self.uuid, self.address, self.name, self.labels)
+
+
+_EMPTY_SNAPSHOT = _MemberListSnapshot(-1, OrderedDict())
+_INITIAL_MEMBERS_TIMEOUT_SECONDS = 120
+
+
+class ClusterService(object):
+    """Cluster service for Hazelcast clients.
+    Allows to retrieve Hazelcast members of the cluster, e.g. by their Address or UUID.
     """
     logger = logging.getLogger("HazelcastClient.ClusterService")
 
-    def __init__(self, config, client, address_providers):
-        self._config = config
+    def __init__(self, client):
         self._client = client
-        self._logger_extras = {"client_name": client.name, "group_name": config.group_config.name}
-        self._members = {}
-        self.owner_connection_address = None
-        self.owner_uuid = None
-        self.uuid = None
-        self.listeners = {}
-
-        for listener in config.membership_listeners:
+        config = client.config
+        self._logger_extras = {"client_name": client.name, "cluster_name": config.cluster_name}
+        self._labels = frozenset(config.labels)
+        self._connection_manager = client.connection_manager
+        self._listeners = {}
+        self._member_list_snapshot = _EMPTY_SNAPSHOT
+        self._initial_list_fetched = threading.Event()
+
+    def start(self, membership_listeners):
```

Comment:
Is this leak to public API?

Reply:
Yes, that should be a private API but since we are using this inside the `client.start`, I made this public. The convention that the user should follow is, if it is not documented, don't use it.

We could make this protected and still be able to call it inside the client module but that could annoy the linters(we don't have any right now, but I am planning to integrate `pylint` and `black` later) due to access to the protected members. Anyways, we could disable that check for this access or remove the method altogether. There are other places in which I did something similar to this. So, I am going to create an issue to track that. There are still to many moving parts, so I am planning to go over it once the structure of the client stabilizes a little bit. Is that OK?

Reply:
>  The convention that the user should follow is, if it is not documented, don't use it.
If this is a well-known convention for Python users, then it is ok. 
Still does not feel good, since what happens if we want to document something but it is a private API.